Userland/crash.cpp

```diff
@@ -1,19 +1,65 @@
+#include <AK/Function.h>
 #include <AK/String.h>
 #include <Kernel/Syscall.h>
 #include <stdio.h>
 #include <stdlib.h>
 #include <sys/mman.h>
 
+#pragma GCC optimize("O0")
+
 static void print_usage_and_exit()
 {
-    printf("usage: crash -[sdiamfMFTtSxyX]\n");
+    printf("usage: crash -[AsdiamfMFTtSxyX]\n");
     exit(0);
 }
 
-#pragma GCC optimize("O0")
+enum CrashFailure {};
+
+class Crash {
+public:
+    Crash(String test_type, Function<CrashFailure()> test)
+        : m_type(test_type)
+        , m_test(move(test))
+    {
+    }
+
+    enum RunType {
+        UsingChildProcess,
+        UsingCurrentProcess,
+    };
+
+    void run(RunType run_type)
+    {
+        if (run_type == RunType::UsingCurrentProcess)
+            m_test();
+
+        printf("\x1B[33mTesting\x1B[0m: \"%s\"\n", m_type.characters());
+
+        // Run the test in a child process so that we do not crash the crash program :^)
+        pid_t pid = fork();
+        if (pid == 0) {
+            m_test();
+            exit(0);
+        }
+
+        // NOTE: We assume that a test passes if it does not exit with an error code of 0
+        int ret;
+        waitpid(pid, &ret, 0);
+        if (ret == 0)
```

Comment:
Assuming the kernel implements the bits for WIFSIGNALED and WSTOPSIG it might be useful to print them out here